### ISSUE TITLE
SSLException when playing a video on Android < 5.0

### ISSUE THREAD

**Issue by ebonhart** [nexxtv/nexxPLAY-android#22]
A while ago, videos stopped playing in our App on Android 4.1 to 4.4. When clicking a video, I see an endless loader, while in the logs I get an SSLException:

**`09-17 13:54:20.398 7714-7929/? W/System.err: javax.net.ssl.SSLException: Connection closed by peer
09-17 13:54:20.403 7714-7929/? W/System.err:     at org.apache.harmony.xnet.provider.jsse.NativeCrypto.SSL_do_handshake(Native Method)
        at org.apache.harmony.xnet.provider.jsse.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:398)
        at libcore.net.http.HttpConnection.setupSecureSocket(HttpConnection.java:209)
        at libcore.net.http.HttpsURLConnectionImpl$HttpsEngine.makeSslConnection(HttpsURLConnectionImpl.java:478)
        at libcore.net.http.HttpsURLConnectionImpl$HttpsEngine.connect(HttpsURLConnectionImpl.java:442)
        at libcore.net.http.HttpEngine.sendSocketRequest(HttpEngine.java:290)
        at libcore.net.http.HttpEngine.sendRequest(HttpEngine.java:240)
        at libcore.net.http.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:282)
        at libcore.net.http.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:497)
09-17 13:54:20.408 7714-7929/? W/System.err:     at libcore.net.http.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:134)
        at tv.nexx.android.player.apiv3.transactions.APICallAsyncTask.doInBackground(APICallAsyncTask.java:62)
        at android.os.AsyncTask$2.call(AsyncTask.java:287)
        at java.util.concurrent.FutureTask.run(FutureTask.java:234)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:230)
09-17 13:54:20.413 7714-7929/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
        at java.lang.Thread.run(Thread.java:841)`**

It should me mentioned that this is not a problem of the newer Nexx player versions. This also occurs in our live version, where Nexx player version 2.1.1 is currently still active. For version 2.1.1, the stacktrace looks a bit different:

**`09-13 16:17:50.400 22499-22526/? W/System.err: javax.net.ssl.SSLPeerUnverifiedException: No peer certificate
        at com.android.org.conscrypt.SSLSessionImpl.getPeerCertificates(SSLSessionImpl.java:146)
        at org.apache.http.conn.ssl.AbstractVerifier.verify(AbstractVerifier.java:93)
        at org.apache.http.conn.ssl.SSLSocketFactory.createSocket(SSLSocketFactory.java:388)
        at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:214)
        at org.apache.http.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:167)
        at org.apache.http.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:125)`**

I did a bit of research and the reason might be that TLSv1 got disabled on the server. When I add the following line to our application class

**`ProviderInstaller.installIfNeeded(getApplicationContext());`**

the endless loader and the SSL error is gone and I get to see advertisements at least. Afterwards, the App crashes with the following exception, however:

**`09-18 11:44:47.350 16181-16181/? W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0x41c92700)
09-18 11:44:47.650 16181-16460/? D/FA: Logging event (FE): app_exception(_ae), Bundle[{timestamp=1537263887361, firebase_event_origin(_o)=clx, firebase_screen_id(_si)=7790696728317143698, firebase_screen_class(_sc)=NexxPlayerActivity, _r=1, fatal=1}]
09-18 11:44:47.695 16181-16181/? E/AndroidRuntime: FATAL EXCEPTION: main
    java.lang.IllegalArgumentException: MediaButtonReceiver component may not be null.
        at android.support.v4.media.session.MediaSessionCompat$MediaSessionImplBase.<init>(MediaSessionCompat.java:1825)
        at android.support.v4.media.session.MediaSessionCompat$MediaSessionImplApi18.<init>(MediaSessionCompat.java:2979)
        at android.support.v4.media.session.MediaSessionCompat.<init>(MediaSessionCompat.java:402)
        at android.support.v4.media.session.MediaSessionCompat.<init>(MediaSessionCompat.java:346)
        at tv.nexx.android.player.ExoPlayerAdapter.preparePlayer(ExoPlayerAdapter.java:205)
        at tv.nexx.android.player.ExoPlayerAdapter.prepare(ExoPlayerAdapter.java:518)
        at tv.nexx.android.player.ExoPlayerAdapter.prepare(ExoPlayerAdapter.java:528)
        at tv.nexx.android.player.logic.Sequencer$MediaState.onEntry(Sequencer.java:425)
        at tv.nexx.android.player.logic.Sequencer.doTransition(Sequencer.java:147)
        at tv.nexx.android.player.logic.Sequencer.startMedia(Sequencer.java:82)
        at tv.nexx.android.player.logic.PlayerLogicStateMachine.startMediaOrGetReady(PlayerLogicStateMachine.java:1987)
        at tv.nexx.android.player.logic.PlayerLogicStateMachine.imaVideoAdEnded(PlayerLogicStateMachine.java:2066)
        at tv.nexx.android.player.logic.IMAManager.onAdEvent(IMAManager.java:146)
        at com.google.obf.ir.a(IMASDK:65)
        at com.google.obf.ir.a(IMASDK:61)
        at com.google.obf.ir.a(IMASDK:98)
        at com.google.obf.ij.a(IMASDK:60)
        at com.google.obf.jd.f(IMASDK:180)
        at com.google.obf.jd.a(IMASDK:44)
        at com.google.obf.je.b(IMASDK:39)
        at com.google.obf.je$1.shouldOverrideUrlLoading(IMASDK:4)
        at android.webkit.CallbackProxy.uiOverrideUrlLoading(CallbackProxy.java:354)
        at android.webkit.CallbackProxy.handleMessage(CallbackProxy.java:488)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:176)
        at android.app.ActivityThread.main(ActivityThread.java:5419)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:525)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
        at dalvik.system.NativeStart.main(Native Method)`**

I was able to get rid of this exception by adding the follwing in the manifest:

**`<receiver android:name="android.support.v4.media.session.MediaButtonReceiver" >
            <intent-filter>
                <action android:name="android.intent.action.MEDIA_BUTTON" />
            </intent-filter>
        </receiver>`**

But this just gets me another exception and App crash after the advertisements:

**`09-18 11:49:08.470 17415-17658/? D/FA: Logging event (FE): app_exception(_ae), Bundle[{timestamp=1537264148155, firebase_event_origin(_o)=clx, firebase_screen_id(_si)=7074289639010539498, firebase_screen_class(_sc)=NexxPlayerActivity, _r=1, fatal=1}]
09-18 11:49:08.495 17415-17415/? E/AndroidRuntime: FATAL EXCEPTION: main
    java.lang.NoClassDefFoundError: android.util.ArrayMap
        at tv.nexx.android.player.InternalUtils.getRunningActivity(InternalUtils.java:124)
        at tv.nexx.android.player.ExoPlayerAdapter.preparePlayer(ExoPlayerAdapter.java:213)
        at tv.nexx.android.player.ExoPlayerAdapter.prepare(ExoPlayerAdapter.java:518)
        at tv.nexx.android.player.ExoPlayerAdapter.prepare(ExoPlayerAdapter.java:528)
        at tv.nexx.android.player.logic.Sequencer$MediaState.onEntry(Sequencer.java:425)
        at tv.nexx.android.player.logic.Sequencer.doTransition(Sequencer.java:147)
        at tv.nexx.android.player.logic.Sequencer.startMedia(Sequencer.java:82)
        at tv.nexx.android.player.logic.PlayerLogicStateMachine.startMediaOrGetReady(PlayerLogicStateMachine.java:1987)
        at tv.nexx.android.player.logic.PlayerLogicStateMachine.imaVideoAdEnded(PlayerLogicStateMachine.java:2066)
        at tv.nexx.android.player.logic.IMAManager.onAdEvent(IMAManager.java:146)
        at com.google.obf.ir.a(IMASDK:65)
        at com.google.obf.ir.a(IMASDK:61)
        at com.google.obf.ir.a(IMASDK:98)
        at com.google.obf.ij.a(IMASDK:60)
        at com.google.obf.jd.f(IMASDK:180)
        at com.google.obf.jd.a(IMASDK:44)
        at com.google.obf.je.b(IMASDK:39)
        at com.google.obf.je$1.shouldOverrideUrlLoading(IMASDK:4)
        at android.webkit.CallbackProxy.uiOverrideUrlLoading(CallbackProxy.java:354)
        at android.webkit.CallbackProxy.handleMessage(CallbackProxy.java:488)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:176)
        at android.app.ActivityThread.main(ActivityThread.java:5419)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:525)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
        at dalvik.system.NativeStart.main(Native Method)`**

So this is where I am stuck. I think the last error might be due to InternalUtils using android.util.ArrayMap, which wasn't introduced until Android 4.4, instead of android.support.v4.util.ArrayMap. 

Long story short: Videos should play for Android 4.1 - 4.4 devices. I don't know if this could be fixed server-side or if you can do anything in the library. I tried most of the steps above on a 4.3 device today, so I coulnd't check if the last step fixed videos for Android 4.4 at least. I can try this on a 4.4 device shortly, however. In the meantime, maybe you can look into this or give your thoughts to begin with.

**Comment by mfentross** [comment]
Hi @ebonhart ,

there is a fix for devices below Lollipop. The version is 3.2.22. Could you please check again with that build?

Cheers

**Comment by ebonhart** [comment]
Unfortunately, the version still has the errors for me.

About the solution I described above: I now tested it with an Android 4.4 device and I was able to play the vidoes that way. So, for Android 4.4 I now have a solution. I would say either you look into your solution again - or, fixing the ArrayMap issue described above might also fix it for Android 4.1 - 4.3
